src/v/cluster/tx_registry_stm.cc

```diff
@@ -0,0 +1,218 @@
+tx_registry_stm::do_sync(model::timeout_clock::duration timeout) {
+    if (!_c->is_leader()) {
+        co_return errc::not_leader;
+    }
+
+    auto ready = co_await persisted_stm::sync(timeout);
+    if (!ready) {
+        co_return errc::generic_tx_error;
+    }
+    co_return _insync_term;
+}
+
+ss::future<> tx_registry_stm::apply(model::record_batch b) {
+    const auto& hdr = b.header();
+
+    if (hdr.type == model::record_batch_type::tx_registry) {
+        vlog(
+          txlog.trace, "processing tx_registry batch at {}", hdr.base_offset);
+
+        vassert(
+          b.record_count() == 1,
+          "tx_registry batch must contain a single record");
+        auto r = b.copy_records();
+        auto& record = *r.begin();
+        auto key = record.release_key();
+
+        auto subtype = serde::from_iobuf<int32_t>(std::move(key));
+        if (subtype == static_cast<int32_t>(batch_subtype::tx_mapping)) {
+            vlog(
+              txlog.trace,
+              "processing tx_registry/tx_mapping cmd at {}",
+              hdr.base_offset);
+            auto value = record.release_value();
+            _mapping = serde::from_iobuf<tx_mapping>(std::move(value));
+            _inited = true;
+        } else {
+            vlog(
+              txlog.trace,
+              "unknown tx_registry/{} cmd at {}",
+              subtype,
+              hdr.base_offset);
+            _seen_unknown_batch_subtype = true;
+        }
+    }
+
+    _insync_offset = b.last_offset();
+
+    _processed++;
+    if (_processed > _log_capacity) {
+        ssx::spawn_with_gate(_gate, [this] { return truncate_log_prefix(); });
+    }
+
+    return ss::now();
+}
+
+std::optional<model::partition_id>
+tx_registry_stm::find_hosting_partition(kafka::transactional_id tid) {
+    for (auto& [partition, hosted] : _mapping.mapping) {
+        if (hosted_transactions::contains(hosted, tid)) {
+            return partition;
+        }
+    }
+    vlog(
+      txlog.error,
+      "tx_registry must cover full tx.id space but {} isn't found",
+      tid);
+    return std::nullopt;
+}
+
+ss::future<bool> tx_registry_stm::try_init_mapping(
+  model::term_id term, int32_t tx_coordinator_partition_amount) {
+    if (_inited) {
+        co_return true;
+    }
+
+    auto units = co_await write_lock();
+    if (_inited) {
+        co_return true;
+    }
+
+    tx_mapping mapping;
+    mapping.id = repartitioning_id(0);
+
+    for (int pid = 0; pid < tx_coordinator_partition_amount; pid++) {
+        model::partition_id partition = model::partition_id(pid);
+        auto initial_hash_range = default_hash_range(
+          partition, tx_coordinator_partition_amount);
+        hosted_txs initial_hosted_transactions{};
+        auto res = hosted_transactions::add_range(
+          initial_hosted_transactions, initial_hash_range);
+        vassert(
+          res == tx_hash_ranges_errc::success,
+          "default txn hash space must be complete");
```

Comment:
would you want to dump out a lot of context here for debugging, considering this is a hard assert anyway?